### PR TITLE
Input for api calls

### PR DESCRIPTION
**Pull request by tomfelix**

### CODE REVIEW

File: src/utils/components/ModalWindow.tsx [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import styled from 'styled-components';
+
+export const ModalWindow = styled.div`
+display: flex;
+flex-direction: column;
+justify-content: space-around;
+align-items: center;
```
Comment (przemyslawjanpietrzak) [inline src/utils/components/ModalWindow.tsx]: as above

File: src/utils/createAndUnlockWeb3.ts [hunk 2]
```diff
@@ -2,21 +2,29 @@ import * as Web3 from 'web3';
...
-    web3 = new Web3(provider);
-    // @ts-ignore
-    window.ethereum.enable().then((account: any) => {
-        const defaultAccount = account[0];
-        web3.eth.defaultAccount = defaultAccount;
-    });
+if (typeof window.ethereum !== 'undefined' || (typeof window.web3 !== 'undefined')) {
+    try {
+        // @ts-ignore
+        const provider = window.ethereum || window.web3.currentProvider;
```
Comment (przemyslawjanpietrzak) [inline src/utils/createAndUnlockWeb3.ts]: why?

File: src/utils/createAndUnlockWeb3.ts [hunk 3]
```diff
@@ -2,21 +2,29 @@ import * as Web3 from 'web3';
...
+        // @ts-ignore
+        web3 = new Web3(provider);
```
Comment (przemyslawjanpietrzak) [inline src/utils/createAndUnlockWeb3.ts]: why?

File: src/utils/createAndUnlockWeb3.ts [hunk 4]
```diff
@@ -2,21 +2,29 @@ import * as Web3 from 'web3';
...
+        // @ts-ignore
+        window.ethereum.enable().then((account: any) => {
```
Comment (przemyslawjanpietrzak) [inline src/utils/createAndUnlockWeb3.ts]: replace any

File: src/App.tsx [hunk 5]
```diff
@@ -1,37 +1,112 @@
...
+      url,
+    };
+    this.setState({
+      requests: {
+        ...requests,
+        [newRequest.hash]: newRequest,
+      },
+    });
+  }
+  handleUpdateState = (updatedState: any) => {
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/App.tsx]: done

File: src/App.tsx [hunk 6]
```diff
@@ -1,37 +1,112 @@
...
+      },
+    });
+  }
+  toggleModal = () => {
+    this.setState({
+      isModalOpen: !this.state.isModalOpen,
+    });
+  }
+
+  handleModal = (modalState: boolean, modalMessage: any) => {
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/App.tsx]: done

File: src/App.tsx [hunk 7]
```diff
@@ -1,37 +1,112 @@
...
+    this.setState({
+      requests: {
+        ...requests,
+        [updatedState.id]: updatedState,
+      },
+    });
+  }
+  toggleModal = () => {
+    this.setState({
+      isModalOpen: !this.state.isModalOpen,
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/App.tsx]: done

File: src/components/AppHeaderNews.tsx [hunk 8]
```diff
@@ -1,7 +1,8 @@
 import styled from 'styled-components';
 
 export const AppHeaderNews = styled.div`
-    display: flex;
+    /* display: flex; */
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/components/AppHeaderNews.tsx]: done

File: src/components/AppHeaderProof.tsx [hunk 9]
```diff
@@ -1,7 +1,8 @@
 import styled from 'styled-components';
 
 export const AppHeaderProof = styled.div`
-    display: flex;
+    /* display: flex; */
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/components/AppHeaderProof.tsx]: done

File: src/customerCalls/CallForm.tsx [hunk 10]
```diff
@@ -1,45 +1,103 @@
...
+    handleTransactionHashAndUrl: (arg1: string, arg2: string) => void;
+    handleModal: (arg1: boolean, arg2: any) => void;
+}
 export const usingOracleContract = new web3.eth.Contract(
     usingOracleAbi,
     process.env.REACT_APP_USING_ORACLE_ADDRESS,
 );
 
-export default class CallForm extends PureComponent<State> {
+export default class CallForm extends PureComponent<Props, State> {
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: changed

File: src/customerCalls/CallForm.tsx [hunk 11]
```diff
@@ -1,45 +1,103 @@
...
     static defaultProps = { query: '' };
     state: State = {
         query: '',
+        networkType: '',
     };
-
+    componentDidMount() {
+        web3.eth.net.getNetworkType()
+            .then((result: any) => {
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: fixed

File: src/customerCalls/CallForm.tsx [hunk 12]
```diff
@@ -1,45 +1,103 @@
...
+                this.setState({
+                    networkType: result,
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: added query variable to this.setState function

File: src/customerCalls/CallForm.tsx [hunk 13]
```diff
@@ -1,45 +1,103 @@
...
         });
     }
+    passHashAndUrlToProps = (hash: string, url: string) => {
+        this.props.handleTransactionHashAndUrl(hash, url);
+    }
 
     handleSubmit = () => {
-        if (this.state.query === '') {
-            alert('put valid url for call');
+        if (usingOracleContract.defaultAccount === undefined) {
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: fixed

File: src/customerCalls/CallForm.tsx [hunk 14]
```diff
@@ -1,45 +1,103 @@
...
-            .send();
-        this.setState({
-            query: '',
-        });
+        if (this.state.query === '') {
+            const message = 'Please put a valid url.';
+            this.props.handleModal(true, message);
+            return;
+        }
+        try {
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: I did't find way to handle all errors, so I used that. but if it is unnecessary I will remove that

File: src/customerCalls/CallForm.tsx [hunk 15]
```diff
@@ -1,45 +1,103 @@
...
+            this.props.handleModal(true, message);
             return;
         }
-        usingOracleContract.methods.request(this.state.query)
-            .send();
-        this.setState({
-            query: '',
-        });
+        if (this.state.query === '') {
+            const message = 'Please put a valid url.';
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: fixed

File: src/customerCalls/CallForm.tsx [hunk 16]
```diff
@@ -1,45 +1,103 @@
...
+            const { query } = this.state;
+            usingOracleContract.methods.request(query)
+                .send()
+                .once('transactionHash', (hash: string) => {
+                    this.passHashAndUrlToProps(hash, query);
+                })
+                .on('error', (error: any) => {
+                    let message = '';
+                    if (error.toString().includes('User denied transaction signature.')) {
+                        message = 'User denied transaction signature.';
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: fixed

File: src/customerCalls/CallForm.tsx [hunk 17]
```diff
@@ -1,45 +1,103 @@
...
+                        this.props.handleModal(true, message);
+                        return;
+                    }
+                    if (error.toString().includes('Error: WalletMiddleware - Invalid "from" address.')) {
+                        message = 'Error: WalletMiddleware - Invalid "from" address.';
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: fixed

File: src/customerCalls/CallForm.tsx [hunk 18]
```diff
@@ -1,45 +1,103 @@
...
+                        this.props.handleModal(true, message);
+                        return;
+                    }
+                    if (error.toString().includes('gas')) {
+                        message = 'You are out of gas!';
+                        this.props.handleModal(true, message);
+                        return;
+                    }
+                    message = 'Something went wrong, please try again';
+                    this.props.handleModal(true, message);
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: fixed

File: src/customerCalls/CallForm.tsx [hunk 19]
```diff
@@ -1,45 +1,103 @@
...
+                    return;
+                });
+            this.setState({
+                query: '',
+            });
+        } catch (error) { console.error(error); }
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: try/catch removed

File: src/customerCalls/CallForm.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/customerCalls/CallForm.tsx]: finally removed after comment regarding catch in try/catch

File: src/requests/Request.tsx [hunk 20]
```diff
@@ -12,6 +12,8 @@ export interface RequestProps {
   url: string;
   isOdd: boolean;
   labels: string[];
+  transactionHash: string;
+  handleUpdateState: object;
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/requests/Request.tsx]: changed

File: src/requests/Request.tsx [hunk 21]
```diff
@@ -94,11 +100,73 @@ class Request extends PureComponent<RequestProps> {
...
+              errorCode
+                ?
+                this.codeMapper(errorCode.toString())
+                :
+                <Loader>Loading...</Loader>
             }
           </RequestContent>
         </RequestTableCell>
       </RequestTableRow>
+      // errorCode ?
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/requests/Request.tsx]: removed

File: src/requests/components/RequestTableCell.tsx [hunk 22]
```diff
@@ -4,7 +4,8 @@ export const RequestTableCell = styled.td`
     box-sizing: border-box;
     text-align: center;
     padding: 10px;
-    width: 20%;
+    /* width: 20%; */
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/requests/components/RequestTableCell.tsx]: removed

File: src/utils/components/ModalFooter.tsx [hunk 23]
```diff
@@ -0,0 +1,5 @@
+import styled from 'styled-components';
+
+export const ModalFooter = styled.div`
+align-self: center;
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/utils/components/ModalFooter.tsx]: done

File: src/utils/components/ModalWindow.tsx (hunk 1 above)
Comment (tomfelix) [reply src/utils/components/ModalWindow.tsx]: done

File: src/App.tsx [hunk 24]
```diff
@@ -1,37 +1,112 @@
...
+  requests: {
+    [key: string]: RequestProps,
+  };
+  isModalOpen: boolean;
+  modalMessage: any;
+}
+
+class App extends React.Component {
+
+  state = {
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/App.tsx]: fixed

File: src/App.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/App.tsx]: changed to RequestStatus interface

File: src/requests/RequestList.tsx [hunk 25]
```diff
@@ -1,73 +1,91 @@
...
+    [key: string]: RequestProps,
+  };
 }
 
-class RequestList extends PureComponent<{}, State> {
+interface Props {
+  requests: {
+    [key: string]: RequestProps,
+  };
+  handleUpdateState: any;
```
[the remark this answers is not in the document]
Comment (tomfelix) [reply src/requests/RequestList.tsx]: done

File: src/utils/createAndUnlockWeb3.ts (hunk 2 above)
Comment (tomfelix) [reply src/utils/createAndUnlockWeb3.ts]: fixed

File: src/utils/createAndUnlockWeb3.ts (hunk 3 above)
Comment (tomfelix) [reply src/utils/createAndUnlockWeb3.ts]: fixed

File: src/utils/createAndUnlockWeb3.ts (hunk 4 above)
Comment (tomfelix) [reply src/utils/createAndUnlockWeb3.ts]: fixed

**Pull request merged**